bindings/python/benchmarks/benchmarks.py

```diff
@@ -30,113 +29,192 @@
+    @staticmethod
+    def exercise_table(table):
+        pass
 
     def time_conventional_ndarray(self):
-        collection = db[collection_names[CUR_SIZE]]
+        collection = db.benchmark
         cursor = collection.find()
-        dtype = dtypes[CUR_SIZE]
-
-        if CUR_SIZE == LARGE:
-            np.array([tuple(doc[k] for k in large_doc_keys) for doc in cursor], dtype=dtype)
+        dtype = self.dtypes
+        if "Large" in type(self).__name__:
+            np.array([tuple(doc[k] for k in self.large_doc_keys) for doc in cursor], dtype=dtype)
         else:
             np.array([(doc["x"], doc["y"]) for doc in cursor], dtype=dtype)
 
     def time_to_numpy(self):
-        c = db[collection_names[CUR_SIZE]]
-        schema = schemas[CUR_SIZE]
-        find_numpy_all(c, {}, schema=schema)
+        c = db.benchmark
+        find_numpy_all(c, {}, schema=self.schema)
 
     def time_conventional_pandas(self):
-        collection = db[collection_names[CUR_SIZE]]
-        _ = dtypes[CUR_SIZE]
+        collection = db.benchmark
         cursor = collection.find(projection={"_id": 0})
         _ = pd.DataFrame(list(cursor))
 
     def time_to_pandas(self):
-        c = db[collection_names[CUR_SIZE]]
-        schema = schemas[CUR_SIZE]
-        find_pandas_all(c, {}, schema=schema)
+        c = db.benchmark
+        find_pandas_all(c, {}, schema=self.schema, projection={"_id": 0})
 
     def time_to_arrow(self):
-        c = db[collection_names[CUR_SIZE]]
-        schema = schemas[CUR_SIZE]
-        find_arrow_all(c, {}, schema=schema)
+        c = db.benchmark
+        table = find_arrow_all(c, {}, schema=self.schema)
+        self.exercise_table(table)
+
+    def time_conventional_arrow(self):
+        c = db.benchmark
+        f = list(c.find({}, projection={"_id": 0}))
+        table = pyarrow.Table.from_pylist(f)
+        self.exercise_table(table)
+
+
+class ProfileReadArray(Read):
+    def setup(self):
+        coll = db.benchmark
+        coll.drop()
+        base_dict = collections.OrderedDict(
+            [("x", 1), ("y", math.pi), ("emb", [math.pi for _ in range(64)])]
+        )
+        schema_dict = {"x": pyarrow.int64(), "y": pyarrow.float64()}
+        dtypes_list = np.dtype([("x", np.int64), ("y", np.float64)])
+        self.schema = Schema(schema_dict)
+        self.dtypes = np.dtype(dtypes_list)
+        coll.insert_many([base_dict.copy() for _ in range(N_DOCS)])
+        print(
+            "%d docs, %dk each with %d keys"
+            % (N_DOCS, len(BSON.encode(base_dict)) // 1024, len(base_dict))
+        )
+
+    # We need this because the naive methods don't always convert nested objects.
+    @staticmethod
+    def exercise_table(table):
+        [
+            [[n for n in i.values] if isinstance(i, pyarrow.ListScalar) else i for i in column]
+            for column in table.columns
+        ]
+
+    def time_to_numpy(self):
+        pass
```

Comment:
TODO?

Reply:
Nested arrays are not supported in NumPy or Pandas with our current implementation IIRC.

Reply:
Gotcha, can you please add a comment then?

Reply:
Done.